akka-distributed-data/src/main/scala/akka/cluster/ddata/VersionVector.scala

```diff
@@ -103,12 +119,60 @@ final case class VersionVector private[akka] (
    */
   def increment(node: Cluster): VersionVector = increment(node.selfUniqueAddress)
 
+  def isEmpty: Boolean =
+    (this eq VersionVector.empty) || (versions match {
+      case Right(_)      ⇒ false
+      case Left(treeMap) ⇒ treeMap.isEmpty
+    })
+
+  /**
+   * INTERNAL API
+   */
+  private[akka] def size: Int =
+    versions match {
+      case Left(treeMap) ⇒ treeMap.size
+      case Right(_)      ⇒ 1
+    }
+
   /**
    * INTERNAL API
    * Increment the version for the node passed as argument. Returns a new VersionVector.
    */
-  private[akka] def increment(node: UniqueAddress): VersionVector =
-    copy(versions = versions.updated(node, Timestamp.counter.getAndIncrement()))
+  private[akka] def increment(node: UniqueAddress): VersionVector = {
+    val v = Timestamp.counter.getAndIncrement()
+    val newVersions = versions match {
+      case Left(treeMap) ⇒
+        if (treeMap.isEmpty) Right(node -> v)
+        else Left(treeMap.updated(node, v))
+      case Right(t @ (n, _)) ⇒
+        if (n == node) Right(n -> v)
+        else Left(emptyVersions + t + (node -> v))
+    }
+    VersionVector(versions = newVersions)
+  }
+
+  /**
+   * INTERNAL API
+   */
+  private[akka] def versionAt(node: UniqueAddress): Long =
+    versions match {
+      case Left(treeMap) ⇒ treeMap.get(node) match {
+        case Some(v) ⇒ v
+        case None    ⇒ Timestamp.Zero
+      }
+      case Right((n, v)) ⇒
```

Comment:
no (I checked something similar with javap). pattern matcher is smarter, there is at least some sanity